Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,324 @@
+          .flatMap(DoubleStream::of)
+          .distinct()
+          .sorted()
+          .toArray();
+
+      double median;
+      int size = doubles.length;
+      if (size == 0) {
+        median = 0;
+      } else {
+        if (size % 2 == 0)
+          median = (doubles[size / 2] + doubles[size / 2 - 1]) / 2;
+        else
+          median = doubles[size / 2];
+      }
+      return median;
+
+    }
+
+    /**
+     * Updates the responsibilities matrix using the similarity values and the
+     * current availability values.
+     */
+    @Override
+    protected void updateResponsibilities()
+    {
+      for(int i = 0 ; i < exampleCount; i++) {
+        double[] rowArray = createSumArray(availabilities[i], similarities[i]);
+        MaxResult maxResult = computeMax(rowArray);
+        for (int j = 0; j < exampleCount; j++) {
+          double maxToSubtract = j != maxResult.maxIndex ? maxResult.maxValue : maxResult.secondMaxValue;
+          responsibilities[i][j] = oneMinusDampingFactor * (similarities[i][j] - maxToSubtract) + dampingFactor * responsibilities[i][j];
+        }
+      }
+    }
+
+    private MaxResult computeMax(double[] rowArray)
+    {
+      MaxResult maxResult = new MaxResult();
+      for (int i = 0; i < rowArray.length; i++) {
+        if(rowArray[i] > maxResult.maxValue){
+          maxResult.secondMaxValue = maxResult.maxValue;
+          maxResult.maxValue = rowArray[i];
+          maxResult.maxIndex = i;
+        }else if(rowArray[i] > maxResult.secondMaxValue){
+          maxResult.secondMaxValue = rowArray[i];
+        }
+      }
+      return maxResult;
+    }
+
+
+    private double[] createSumArray(double[] availability, double[] similarity)
+    {
+      double[] resultArray = new double[availability.length];
+      for (int i = 0; i < availability.length; i++) {
+        resultArray[i] = availability[i] + similarity[i];
+      }
+      return resultArray;
+    }
+
+    /**
+     * Updates the availabilities matrix based on the current responsibility
+     * values.
+     */
+    @Override
+    protected void updateAvailabilities()
+    {
+      Vector vector = VectorFactory.getDefault().createVector(exampleCount);
+      for (int j = 0; j < exampleCount; j++) {
+        removeNegativeValues(vector, responsibilities, j);
+        vector.set(j, responsibilities[j][j]);
+        updateAvailabilitiesMatrix(j, vector, availabilities);
+      }
+    }
+
+    private void updateAvailabilitiesMatrix(int j, Vector columnVector, double[][] availabilities)
+    {
+      double oldValue;
+      for (int i = 0; i < availabilities[0].length; i++) {
+        oldValue = availabilities[i][j];
+        double newValue = sum(columnVector) - columnVector.get(i);
+        if(i != j){
+          newValue = Math.min(newValue, 0);
+        }
+        availabilities[i][j] = oneMinusDampingFactor * newValue + dampingFactor * oldValue;
+      }
+    }
+
+    private double sum(Vector columnVector) {

```

Comment:
You should be able to just do columnVector.sum() instead of needing this method.